rust/osm2lanes/src/transform/tags_to_lanes/road.rs

```diff
@@ -0,0 +1,412 @@
+    Guessed(T),
+}
+impl<T> Hunch<T> {
+    fn get_ref(&self) -> &T {
+        match self {
+            Tagged(v) | Implied(v) | Assumed(v) | Guessed(v) => &v,
+        }
+    }
+
+    fn with<U>(&self, val: U) -> Hunch<U> {
+        match self {
+            Tagged(_) => Tagged(val),
+            Implied(_) => Implied(val),
+            Assumed(_) => Assumed(val),
+            Guessed(_) => Guessed(val),
+        }
+    }
+    /// Returns a hunch of the value, with an variation appropriate to combining information from
+    /// two other hunches.
+    fn combine<U, V>(val: T, a: &Hunch<U>, b: &Hunch<V>) -> Hunch<T> {
+        match (a, b) {
+            (Guessed(_), _) | (_, Guessed(_)) => Guessed(val),
+            (Assumed(_), _) | (_, Assumed(_)) => Assumed(val),
+            (Implied(_), _) | (_, Implied(_)) => Implied(val),
+            (Tagged(_), _) | (_, Tagged(_)) => Tagged(val),
+        }
+    }
+    fn downgrade<S: HunchStrength>(self, sources: &Vec<&S>) -> Self {
+        // TODO should implement the sortable trait to make this easier:
+        let mut strength = &self;
+        for s in sources {
+            match (strength, s.strength()) {
+                (Guessed(_), _) | (_, Guessed(_)) => Guessed(()),
+                (Assumed(_), _) | (_, Assumed(_)) => Assumed(()),
+                (Implied(_), _) | (_, Implied(_)) => Implied(()),
+                (Tagged(_), _) | (_, Tagged(_)) => Tagged(()),
+            }
+        }
+        strength.with(self.into_inner())
+    }
+    // fn combine<I, U where I: Iter<&Hunch<U>>>(val: T, hunches: I) {
+    //
+    // }
+    /// Returns the stronger of two hunches (preferring self), emitting an error if there is a conflict.
+    fn reconcile(self: &Hunch<T>, b: &Hunch<T>, on_conflict: fn()) -> Hunch<T> {
+        match (self, b) {
+            (Implied(i), Tagged(t)) | (Tagged(t), Implied(i)) if t != i => {
+                on_conflict();
+                Implied(i) // TODO return self instead?
+            }
+            (Tagged(v), _) | (_, Tagged(v)) => Ok(Tagged(v)),
+            (Implied(v), _) | (_, Implied(v)) => Ok(Implied(v)),
+            (Assumed(v), _) | (_, Assumed(v)) => Ok(Assumed(v)),
+            (Guessed(v), _) | (_, Guessed(v)) => Ok(Guessed(v)),
+        }
+    }
+}
+
+trait HunchStrength {
+    fn strength(&self) -> Hunch<()>;
+}
+impl HunchStrength for Hunch<_> {
+    fn strength(&self) -> Hunch<()> {
+        match self {
+            Tagged(_) => Tagged(()),
+            Implied(_) => Implied(()),
+            Assumed(_) => Assumed(()),
+            Guessed(_) => Guessed(()),
+        }
+    }
+}
+impl HunchStrength for Theory<_> {
+    fn strength(&self) -> Hunch<()> {
+        match self.hunch() {
+            Some(h) => h.strength(),
+            None => Guessed(()),
+        }
+    }
+}
+
+// trait HasProblems { fn iter<Problem>... }
+// type TheoryHistory = Vec<Problem>;
+struct Theory<'a, T> {
+    history: Vec<(Hunch<T>, Vec<&'a Theory<'a, T>>)>,
```

Comment:
It's almost as if we're producing a full proof of how we arrive at the final answer and all of the assumptions made along the way. Really interesting idea, I've never seen something like this before. Interested to see where this approach goes

Reply:
Yes, it just feels silly to throw any of the history away at any of the points where I have the option to. Iterating over the whole set of problems seems like the first useful reason to have any history around.

This line might be the reason I put in the lifetime. I am trying to keep a reference to the `Theory`/`impl TheoryLike` thing in the history, so we don't copy...